src/fdopendir_support.c

```diff
@@ -0,0 +1,95 @@
+
+/*
+ * Copyright (c) 2023 raf <raf@raf.org>
+ *
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/* MP support header */
+#include "MacportsLegacySupport.h"
+#if __MP_LEGACY_SUPPORT_FDOPENDIR__
+
+#include "common-priv.h"
+
+#define __MP_LEGACY_SUPPORT_NO_DIRFD_MACRO
+
+#include <stdlib.h>
+#include <dirent.h>
+#include <sys/errno.h>
+
+
+/*
+ * Wrapped version of opendir() for fdopendir() compatibility
+ *
+ * On success, this function returns allocated memory that must be
+ * deallocated by __mpls_closedir() (see closedir macro in dirent.h).
+ */
+
+DIR *__mpls_opendir(const char *name) {
+    void *dirvp; /* Real DIR ptr as a void ptr */
+    DIR *dirp; /* Real DIR ptr plus supplied directory file descriptor */
+
+    dirvp = (opendir)(name);
+    if (!dirvp) {
+        return 0;
+    }
+
+    dirp = (DIR *)malloc(sizeof(DIR));
+    if (!dirp) {
+        (void)(closedir)(dirvp);
+        errno = ENOMEM;
+        return 0;
+    }
+
+    dirp->__mpls_dirp = dirvp;
+    dirp->__mpls_dirfd = -1;
+
+    return dirp;
+}
+
+/*
+ * Wrapped version of closedir() for fdopendir() compatibility (see closedir
+ * macro in dirent.h).
+ *
+ * This function deallocates memory that was allocated by fdopendir() or
+ * __mpls_opendir().
+ */
+
+int __mpls_closedir(DIR *dirp) {
+    if (!dirp) {
+        errno = EBADF;
+        return -1;
+    }
+
+    int rc = (closedir)(dirp->__mpls_dirp);
+
+    if (dirp->__mpls_dirfd != AT_FDCWD && dirp->__mpls_dirfd != -1)
+        PROTECT_ERRNO(close(dirp->__mpls_dirfd));
+
+    free(dirp);
+
+    return rc;
+}
+
+/*
+ * Wrapped version of dirfd() for fdopendir() compatibility because dirfd()
+ * itself is already a macro (see dirfd macro in dirent.h).
+ */
+
+int __mpls_dirfd(DIR *dirp)
+{
+    return dirfd(dirp->__mpls_dirp);
```

Comment:
To avoid the confusion of `dirfd(fdopendir(fd)) != fd`, maybe this should do
```c
if (dirp->__mpls_dirfd >= 0) {
    return dirp->__mpls_dirfd;
}
```

Reply:
Thanks. It's done. I was wondering which fd would be better to return. Since whatever calls dirfd() isn't allowed to change the state, and both refer to the same directory, I thought it shouldn't really matter which. But if once choice can lead to confusion, then the other choice is better.